<!DOCTYPE html> <html lang="en"> <head> <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"> <meta charset="utf-8"> <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"> <meta http-equiv="X-UA-Compatible" content="IE=edge"> <title> Familial breast cancer | Héctor Climente-González </title> <meta name="author" content="Héctor Climente-González"> <meta name="description" content="Molecular basis of familial breast cancer"> <meta name="keywords" content="machine-learning, genetics, pharma, drug-discovery, climente, climente-gonzalez"> <link rel="stylesheet" href="/assets/css/bootstrap.min.css?a4b3f509e79c54a512b890d73235ef04"> <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/css/mdb.min.css" integrity="sha256-jpjYvU3G3N6nrrBwXJoVEYI/0zw8htfFnhT9ljN3JJw=" crossorigin="anonymous"> <link defer rel="stylesheet" href="/assets/css/academicons.min.css?f0b7046b84e425c55f3463ac249818f5"> <link defer rel="stylesheet" href="/assets/css/scholar-icons.css?62b2ac103a88034e6882a5be5f3e2772"> <link defer rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700|Roboto+Slab:100,300,400,500,700|Material+Icons&amp;display=swap"> <link defer rel="stylesheet" href="/assets/css/jekyll-pygments-themes-github.css?591dab5a4e56573bf4ef7fd332894c99" media="" id="highlight_theme_light"> <link defer href="/assets/css/bootstrap-toc.min.css?6f5af0bb9aab25d79b2448143cbeaa88" rel="stylesheet"> <link rel="shortcut icon" href="data:image/svg+xml,&lt;svg%20xmlns=%22http://www.w3.org/2000/svg%22%20viewBox=%220%200%20100%20100%22&gt;&lt;text%20y=%22.9em%22%20font-size=%2290%22&gt;%F0%9F%AA%A2&lt;/text&gt;&lt;/svg&gt;"> <link rel="stylesheet" href="/assets/css/main.css?d41d8cd98f00b204e9800998ecf8427e"> <link rel="canonical" href="https://hclimente.github.io/blog/familial-breast-cancer/"> <script src="/assets/js/theme.js?9a0c749ec5240d9cda97bc72359a72c0"></script> <link defer rel="stylesheet" href="/assets/css/jekyll-pygments-themes-native.css?5847e5ed4a4568527aa6cfab446049ca" media="none" id="highlight_theme_dark"> <script>
    initTheme();
  </script> </head> <body class="fixed-top-nav sticky-bottom-footer"> <header> <nav id="navbar" class="navbar navbar-light navbar-expand-sm fixed-top" role="navigation"> <div class="container"> <a class="navbar-brand title font-weight-lighter" href="/"> <span class="font-weight-bold">Héctor</span> Climente-González </a> <button class="navbar-toggler collapsed ml-auto" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar top-bar"></span> <span class="icon-bar middle-bar"></span> <span class="icon-bar bottom-bar"></span> </button> <div class="collapse navbar-collapse text-right" id="navbarNav"> <ul class="navbar-nav ml-auto flex-nowrap"> <li class="nav-item "> <a class="nav-link" href="/">about </a> </li> <li class="nav-item active"> <a class="nav-link" href="/blog/">blog </a> </li> <li class="nav-item "> <a class="nav-link" href="/publications/">publications </a> </li> <li class="nav-item "> <a class="nav-link" href="/repositories/">repositories </a> </li> <li class="nav-item "> <a class="nav-link" href="/cv/">cv </a> </li> <li class="nav-item"> <button id="search-toggle" title="Search" onclick="openSearchModal()"> <span class="nav-link">ctrl k <i class="ti ti-search"></i></span> </button> </li> <li class="toggle-container"> <button id="light-toggle" title="Change theme"> <i class="ti ti-sun-moon" id="light-toggle-system"></i> <i class="ti ti-moon-filled" id="light-toggle-dark"></i> <i class="ti ti-sun-filled" id="light-toggle-light"></i> </button> </li> </ul> </div> </div> </nav> <progress id="progress" value="0"> <div class="progress-container"> <span class="progress-bar"></span> </div> </progress> </header> <div class="container mt-5" role="main"> <div class="row"> <div class="col-sm-3"> <nav id="toc-sidebar" class="sticky-top"></nav> </div> <div class="col-sm-9"> <div class="post"> <header class="post-header"> <h1 class="post-title">Familial breast cancer</h1> <p class="post-meta"> Created in July 20, 2017 </p> <p class="post-tags"> <a href="/blog/2017"> <i class="fa-solid fa-calendar fa-sm"></i> 2017 </a>   ·   <a href="/blog/tag/cancer"> <i class="fa-solid fa-hashtag fa-sm"></i> cancer</a>   <a href="/blog/tag/genetics"> <i class="fa-solid fa-hashtag fa-sm"></i> genetics</a> </p> </header> <article class="post-content"> <div id="markdown-content"> <h1 id="cancer">Cancer</h1> <p>Cancer is the name of a collection of related diseases. Specifically, all cancers undergo an uncontrolled proliferation of the patient cells, which spread into surrounding tissues. In a normal organism, cells grow and divide to maintain the tissue. As cells grow old, or accumulate too much damage, they undergo cell death and new cells will take their place. However, in cancer, this orderly process breaks down. Cells refuse to die when they get old, or accumulate damage. New cells are formed even if they are not needed. In consequence they form purpose-less growths called tumors.</p> <p>This abnormal behavior occurs as consequence of the alteration of crucial genes. These alterations can be inherited from our parents, or acquired during our lifetime due to replication errors or exposure to DNA-damaging substances. As with any other phenotypic trait, the likelihood of developing cancer will be determined by the interplay between our genetic background and the environment: genetic backgrounds may favor or hinder the acquisition of mutations, and so do environmental factors.</p> <h1 id="breast-cancer">Breast cancer</h1> <p>Breast cancer occurs when breast cells undergo this uncontrolled proliferation. In most of the cases they begin in the ducts that carry the milk to the nipple. However the tumor can originate in other tissues, mainly the milk-producing gland.</p> <p>Breast cancer is the second most commonly diagnosed cancer among women, after non-melanoma skin cancer. We find the highest incidence in Western countries, and it has increased in the last decades. Two proposed explanations are more diagnoses due to mammography screening, or more cases due to the generalization of hormone treatments for menopause. It is also the second leading cause of cancer deaths after lung cancer. Encouragingly, the prognosis is better every year (5y survival rate &gt; 80% in France) thanks to improvements in treatments and better screening. Unsurprisingly, the mortality is higher in developing countries. It is mostly a women’s disease: only about 1% of the diagnosed cases are in men. Among the most important risk factors for breast cancer we can highlight age, family history, reproductive history, usage of oral contraceptives, overweight, physical activity and exposure to radiation.</p> <h1 id="breast-cancer-subtypes">Breast cancer subtypes</h1> <p>Breast cancer is a very heterogeneous disease: while all the tumors appear in the same organ, the tissue where they originate, the molecular mechanism involved, the response to therapy, etc. vastly differ. In general, clinical decisions are based on the expression of 3 molecular markers: the expression of the endocrine receptors for estrogen and progesterone (ER and PgR, respectively) and the overexpression of the HER2 gene. The proteins these three genes code for are targets for chemotherapy. Based on the results, we distinguish three main breast cancer subtypes: hormone receptor positive, HER2 positive and triple negative.</p> <h2 id="hormone-receptor-positive">Hormone receptor positive</h2> <p>Hormone receptor positive tumors include the tumors expressing ER and/or PR, which respectively depend on estrogen and/or progesterone to grow. They happen mostly in postmenopausal women. HR+/HER2- also known as LuminalA are the majority of breast cancers (60-75%) and they present the best prognosis.</p> <h2 id="her2-positive">HER2 positive</h2> <p>HER2+ tumors depend on the protein HER2/neu (human epidermal growth factor receptor 2) to proliferate, which they over-express. HR+/HER2+ (also known as LuminalB) constitutes 10% of the cases, while HR-/HER2 (also known as HER2-enriched) involves 5% of them. There are a couple of very effective drugs against it.</p> <h2 id="triple-negative">Triple-negative</h2> <p>Triple-negative tumors (also known as basal-like) lack the expression of all three of ER, PgR and HER2. These patients present a worse prognosis than the rest, due to the aggressiveness of the tumor and the lack of a clear molecular target. Still, the main treatment is chemotherapy.</p> <h1 id="screening">Screening</h1> <p>Routine screening for breast cancer has been set up in the last decades in many countries. After it was introduced we observed a mortality reduction of 30%, justifying its wide implementation. It consists on yearly mammographies. This technique uses low resolution X-rays to detect lesions in the breast. If the lesion is big enough, further morphological examination allows to characterize if the lesion is benign or malign. However, most of the screen-detected lesions are small. In consequence, we perform an additional intervention (biopsy or advanced imaging technique) to make the diagnosis.</p> <p>Despite the aforementioned success, there is a big controversy regarding mammography programs usefulness. That is because, in practical terms, the 30% decrease in mortality means that for every 1000 women examined, we will save 7-9 lives. On the other hand, it will provoke with 4 cases of overdiagnosis, tumors that would have never progressed to symptomatic presentation during the lifetime of the woman. That, and many more cases which will require further, possibly invasive, testing to make the final (negative) diagnosis. The proposed solution to this involves adapting the screening process to different demographic groups, so that high risk groups are closely watched while low risk groups loosely so.</p> <h1 id="familial-breast-cancer">Familial breast cancer</h1> <p>In the mid-19th century a French medical doctor, Pierre Paul Broca, reported for the first time a case of familial breast cancer. Indeed, his wife acquired breast cancer, as many women in her family had for, at least, 4 generations. Cases of familiar breast cancer usually occur in women younger than 50 years, and bilateral primary breast tumors are frequent. Epidemiological studies later quantified the relative risk conferred by a the presence of multiple breast cancers in the family at 2.7. Moreover, they exhibit a higher likelihood of acquiring triple-negative breast cancer.</p> <h1 id="genes-involved-in-familial-breast-cancer">Genes involved in familial breast cancer</h1> <p>It wasn’t until the late 20th century that two genes involved in DNA repair, BRCA1 and BRCA2, were associated with hereditary breast and ovarian cancer (HBOC). Some mutations in these genes increase the risk of developing breast cancer, giving respectively a 57–65% or 45–55% risk of developing breast cancer by age 70 among women. For that reason, BRCA1 and BRCA2 mutations are rare in most populations (1 of 400).</p> <p>HBOC follows an autosomal dominant inheritance pattern. While approximately 5–10% of all patients with breast cancer exhibit a monogenic predisposition to breast and ovarian cancer, only about 25% of them harbor BRCA1/2 mutations. Other 23 genes have been associated with familial breast and/or ovarian cancer (Table 1).</p> <p>Table 1. Overview of HBOC genes: estimated lifetime risk of breast cancer (age in years) and tumorogenic molecular mechanisms that involves them: homologous recombination repair (HRR), replication fork stability, transcription–replication collisions, mismatch repair (MMR), DNA damage signaling, checkpoints and cell death, and/or others. Adapted from <em>Nielsen et al.</em></p> <table> <thead> <tr> <th style="text-align: center"><strong>Gene</strong></th> <th style="text-align: center"><strong>Breast cancer estimated lifetime risk (age in years)</strong></th> <th style="text-align: center"><strong>HRR</strong></th> <th style="text-align: center"><strong>Replication fork stability</strong></th> <th style="text-align: center"><strong>Transcription–replication collisions</strong></th> <th style="text-align: center"><strong>MMR</strong></th> <th style="text-align: center"><strong>DNA damage signaling, checkpoints and cell death</strong></th> <th style="text-align: center"><strong>Other</strong></th> </tr> </thead> <tbody> <tr> <td style="text-align: center">ATM</td> <td style="text-align: center">60% by age 80</td> <td style="text-align: center">✓</td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center">✓</td> <td style="text-align: center"> </td> </tr> <tr> <td style="text-align: center">BARD1</td> <td style="text-align: center">Unknown</td> <td style="text-align: center">✓</td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> </tr> <tr> <td style="text-align: center">BLM</td> <td style="text-align: center">Unknown</td> <td style="text-align: center"> </td> <td style="text-align: center">✓</td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> </tr> <tr> <td style="text-align: center">BRCA1</td> <td style="text-align: center">57–65% by age 70</td> <td style="text-align: center">✓</td> <td style="text-align: center">✓</td> <td style="text-align: center">✓</td> <td style="text-align: center"> </td> <td style="text-align: center">✓</td> <td style="text-align: center"> </td> </tr> <tr> <td style="text-align: center">BRCA2</td> <td style="text-align: center">45–55% by age 70</td> <td style="text-align: center">✓</td> <td style="text-align: center">✓</td> <td style="text-align: center">✓</td> <td style="text-align: center"> </td> <td style="text-align: center">✓</td> <td style="text-align: center"> </td> </tr> <tr> <td style="text-align: center">BRIP1</td> <td style="text-align: center">OR: &lt;2.0</td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> </tr> <tr> <td style="text-align: center">CDH1</td> <td style="text-align: center">42% by age 80</td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center">✓</td> </tr> <tr> <td style="text-align: center">CHEK2</td> <td style="text-align: center">37% by age 70</td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center">✓</td> <td style="text-align: center"> </td> </tr> <tr> <td style="text-align: center">FAM175A</td> <td style="text-align: center">Unknown</td> <td style="text-align: center">✓</td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> </tr> <tr> <td style="text-align: center">FANCC</td> <td style="text-align: center">Unknown</td> <td style="text-align: center"> </td> <td style="text-align: center">✓</td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> </tr> <tr> <td style="text-align: center">FANCM</td> <td style="text-align: center">Unknown</td> <td style="text-align: center"> </td> <td style="text-align: center">✓</td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> </tr> <tr> <td style="text-align: center">MLH1</td> <td style="text-align: center">~19% by age 70</td> <td style="text-align: center"> </td> <td style="text-align: center">✓</td> <td style="text-align: center"> </td> <td style="text-align: center">✓</td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> </tr> <tr> <td style="text-align: center">MRE11</td> <td style="text-align: center">Unknown</td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center">✓</td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> </tr> <tr> <td style="text-align: center">MSH2</td> <td style="text-align: center">~11% by age 70</td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center">✓</td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> </tr> <tr> <td style="text-align: center">NBN</td> <td style="text-align: center">OR: 3.0</td> <td style="text-align: center">✓</td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> </tr> <tr> <td style="text-align: center">NF1</td> <td style="text-align: center">6.5-fold increase in women aged 30–39</td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center">✓</td> </tr> <tr> <td style="text-align: center">PALB2</td> <td style="text-align: center">35% by age 70</td> <td style="text-align: center">✓</td> <td style="text-align: center">✓</td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> </tr> <tr> <td style="text-align: center">PMS2</td> <td style="text-align: center">SIR: 3.8</td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center">✓</td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> </tr> <tr> <td style="text-align: center">PTEN</td> <td style="text-align: center">85% by age 70</td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center">✓</td> </tr> <tr> <td style="text-align: center">RAD51B</td> <td style="text-align: center">Unknown</td> <td style="text-align: center">✓</td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> </tr> <tr> <td style="text-align: center">RAD51C</td> <td style="text-align: center">Unknown</td> <td style="text-align: center">✓</td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> </tr> <tr> <td style="text-align: center">RAD51D</td> <td style="text-align: center">Unknown</td> <td style="text-align: center">✓</td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> </tr> <tr> <td style="text-align: center">RECQL</td> <td style="text-align: center">Unknown</td> <td style="text-align: center"> </td> <td style="text-align: center">✓</td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> </tr> <tr> <td style="text-align: center">RINT1</td> <td style="text-align: center">Unknown</td> <td style="text-align: center">✓</td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> </tr> <tr> <td style="text-align: center">STK11</td> <td style="text-align: center">32% by age 60</td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center">✓</td> </tr> <tr> <td style="text-align: center">TP53</td> <td style="text-align: center">25% by age 70</td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center"> </td> <td style="text-align: center">✓</td> <td style="text-align: center"> </td> </tr> </tbody> </table> <h2 id="molecular-function">Molecular function</h2> <p>Nearly all known HBOC susceptibility genes encode tumor suppressors that participate in genome stability pathways (homologous recombination repair, replication fork stability, transcription–replication collisions, mismatch repair, and DNA damage signaling, checkpoints and cell death).</p> <h3 id="homologous-recombination-repair">Homologous recombination repair</h3> <p>The homologous recombination repair pathway (HRR) deals with double strand DNA breaks by using the undamaged chromosome as template for error-free repair. After a DSB occurs, the MRN complex (MRE11, RAD50 and NDN) detects and binds the free DNA ends. Then, it promotes DNA damage checkpoint signaling.</p> <p>HRR involves BRCA1, BRCA2 and, actually, most of the HBOC genes. Because of its ability to interact with a wide range of proteins, BRCA1 is hypothetized to act as a recruitment scaffold. A deficiency of BRCA1 is linked to the inability to trigger HRR. Mutations in the MRN complex have also been clinically associated to breast cancer, although dubiously so in the case of RAD50 variants. Reassuringly, some other HBOC genes are interactors of the MRN complex and BRCA1/2.</p> <h3 id="replication-fork-stability">Replication fork stability</h3> <p>BRCA1 and BRCA2 protect newly synthesized DNA and promote the restart of stalled forks in an HRR-independent manner. In the absence of these proteins, newly synthesized DNA in a stalled fork would get degraded, leading to genome instability and increasing the risk of cancer.</p> <h3 id="transcriptionreplication-collisions">Transcription–replication collisions</h3> <p>Collisions between transcription and replication are emerging as a source of genome instability. In particular, RNA-DNA hybrids called R-loops can form between the nascent transcript and the DNA template. They can lead to double-strand breaks and mutations. Both BRCA1 and BRCA2 participate in the resolution of R-loops, preventing their accumulation. In consequence, BRCA-deficient cells tend to suffer transcriptional stress that leads to genome instability. Nonetheless the relationship between this mechanism and proneness to HBOC is yet to be proven, and the genes involved further investigated.</p> <h3 id="mismatch-repair">Mismatch repair</h3> <p>DNA mismatch repair (MMR) corrects base-base mispairs. When MMR is faulty, accumulations point mutations and genetic changes in repeated nucleotide sequences (microsatellite instability) occur. MMR also plays a role in error-free HRR.</p> <h3 id="dna-damage-signaling-checkpoints-and-cell-death">DNA damage signaling, checkpoints and cell death</h3> <p>Pathways involved in genome maintenance, cell cycle checkpoints and cell death usually eliminate cells with damaged DNA. When proteins involved in them are not active, some processes such as cell cycle arrest, apoptosis and senescence will not occur. In consequence, cells that undergo genomic alterations are allowed to proliferate. The most famous case of HBOC in this pathway is TP53, which coordinates the transcriptional induction of many genome stability factors.</p> <h1 id="missing-heritability">Missing heritability</h1> <p>Despite the identification of HBOC genes, 52% of the heritability or familial breast cancer remains unexplained: 20% is explained by high penetrance loci and an extra 28% by common variants. We can illustrate this point with the largest genetic association in familial breast cancer so far, carried out in 2013 by Michailidou <em>et al.</em>. The study had two steps. In a first stage, 10,052 breast cancer cases and 12,575 controls of European ancestry were genotyped using the iCOGS platform. Through a GWAS, they selected 29,807 SNPs for further examination in a second stage on a larger cohort (45,290 cases and 41,880 controls). Despite estimating that at least 1,000 uncorrelated loci were involved in breast cancer susceptibility, they were only able to get a genome wide significance (p &lt; 5 × 10<sup>−8</sup>) for ~5% of them (41 SNPs). Moreover, the 41 SNPs have low effect sizes (OR &lt;= 1.26). In fact, only 10 of the +1000 candidate SNPs showed an OR &gt; 1.05.</p> <blockquote> <p><strong>iCOGS</strong> is a custom Illumina array designed by four consortia that study genetic susceptibility of three hormone-related cancers: breast (<a href="http://ccge.medschl.cam.ac.uk/consortia/bcac/" rel="external nofollow noopener" target="_blank">BCAC</a> and <a href="http://ccge.medschl.cam.ac.uk/consortia/cimba/" rel="external nofollow noopener" target="_blank">CIMBA</a>), ovarian (<a href="http://ccge.medschl.cam.ac.uk/consortia/ocac/" rel="external nofollow noopener" target="_blank">OCAC</a>) and prostate (<a href="http://ccge.medschl.cam.ac.uk/consortia/practical/" rel="external nofollow noopener" target="_blank">PRACTICAL</a>). Its explicit purpose is to facile the genotyping in large case-control studies for these tumors. The criteria to include the SNPs was (i) previously associated with cancer susceptibility or survival; (ii) fine mapping of genomic regions of interest; (iii) associated with cancer-related quantitative traits; (iv) in selected candidate genes or pathways; (v) associated with other cancers. The final array included ~211,000 SNPs.</p> </blockquote> <h2 id="genesis">GENESIS</h2> <p>GENESIS (GENE SISter) is a French project that aims to shed some light on familial breast cancer. The index cases are patients with a breast cancer affected sister and no BRCA1/2 mutations. The controls are unaffected colleagues and/or friends of the cases. This dataset is specially interesting for us: the heritability of the disease among the cases is not driven by BRCA1/2, but by rarer variants which are enriched in this experimental setting. GENESIS used the iCOGS genotyping platform.</p> <h1 id="references">References</h1> <ul> <li>PDQ® Cancer Genetics Editorial Board. PDQ Genetics of Breast and Gynecologic Cancers. Bethesda, MD: National Cancer Institute. Updated 30/03/2017. Available at: https://www.cancer.gov/types/breast/hp/breast-ovarian-genetics-pdq. Accessed 14/04/2017. [PMID: 26389210]</li> <li>Foulkes, W. D. (2008). Inherited Susceptibility to Common Cancers. The New England Journal of Medicine, 359(20), 2143–2153. https://doi.org/10.1056/NEJMra0802968</li> <li>Nielsen, F. C., van Overeem Hansen, T., &amp; Sørensen, C. S. (2016). Hereditary breast and ovarian cancer: new genes in confined pathways. Nature Reviews Cancer, 16(9), 599–612. https://doi.org/10.1038/nrc.2016.</li> <li>Michailidou, K., Hall, P., Gonzalez-Neira, A., Ghoussaini, M., Dennis, J., Milne, R. L., … Easton, D. F. (2013). Large-scale genotyping identifies 41 new loci associated with breast cancer risk. Nature Genetics, 45(4), 353–361. https://doi.org/10.1038/ng.2563</li> <li>Rudolph, A., Chang-claude, J., &amp; Schmidt, M. K. (2016). Gene – environment interaction and risk of breast cancer. British Journal of Cancer, 114(2), 125–133. https://doi.org/10.1038/bjc.2015.439</li> <li>Sakoda, L. C., Jorgenson, E., &amp; Witte, J. S. (2013). Turning of COGS moves forward findings for hormonally mediated cancers. Nature Genetics, 45(4), 345–8. https://doi.org/10.1038/ng.2587</li> <li>Sinilnikova, O. M., Dondon, M.-G., Eon-Marchais, S., Damiola, F., Barjhoux, L., Marcou, M., … Andrieu, N. (2016). GENESIS: a French national resource to study the missing heritability of breast cancer. BMC Cancer, 16(1), 13. https://doi.org/10.1186/s12885-015-2028-9</li> </ul> </div> </article> <div id="giscus_thread" style="max-width: 930px; margin: 0 auto;"> <script>
      let giscusTheme = determineComputedTheme();
      let giscusAttributes = {
        src: 'https://giscus.app/client.js',
        'data-repo': 'hclimente/new_website_test',
        'data-repo-id': '',
        'data-category': 'Comments',
        'data-category-id': '',
        'data-mapping': 'title',
        'data-strict': '1',
        'data-reactions-enabled': '1',
        'data-emit-metadata': '0',
        'data-input-position': 'bottom',
        'data-theme': giscusTheme,
        'data-lang': 'en',
        crossorigin: 'anonymous',
        async: '',
      };

      let giscusScript = document.createElement('script');
      Object.entries(giscusAttributes).forEach(([key, value]) => giscusScript.setAttribute(key, value));
      document.getElementById('giscus_thread').appendChild(giscusScript);
    </script> <noscript>Please enable JavaScript to view the <a href="http://giscus.app/?ref_noscript" rel="external nofollow noopener" target="_blank">comments powered by giscus.</a> </noscript> </div> </div> </div> </div> </div> <footer class="sticky-bottom mt-5" role="contentinfo"> <div class="container"> © Copyright 2025 Héctor Climente-González. </div> </footer> <script src="https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js" integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script> <script src="/assets/js/bootstrap.bundle.min.js"></script> <script src="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/js/mdb.min.js" integrity="sha256-NdbiivsvWt7VYCt6hYNT3h/th9vSTL4EDWeGs5SN3DA=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/masonry-layout@4.2.2/dist/masonry.pkgd.min.js" integrity="sha256-Nn1q/fx0H7SNLZMQ5Hw5JLaTRZp0yILA/FRexe19VdI=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/imagesloaded@5.0.0/imagesloaded.pkgd.min.js" integrity="sha256-htrLFfZJ6v5udOG+3kNLINIKh2gvoKqwEhHYfTTMICc=" crossorigin="anonymous"></script> <script defer src="/assets/js/masonry.js?a0db7e5d5c70cc3252b3138b0c91dcaf" type="text/javascript"></script> <script defer src="https://cdn.jsdelivr.net/npm/medium-zoom@1.1.0/dist/medium-zoom.min.js" integrity="sha256-ZgMyDAIYDYGxbcpJcfUnYwNevG/xi9OHKaR/8GK+jWc=" crossorigin="anonymous"></script> <script defer src="/assets/js/zoom.js?85ddb88934d28b74e78031fd54cf8308"></script> <script defer src="/assets/js/bootstrap-toc.min.js?c82ff4de8b0955d6ff14f5b05eed7eb6"></script> <script src="/assets/js/no_defer.js?2781658a0a2b13ed609542042a859126"></script> <script defer src="/assets/js/common.js?e0514a05c5c95ac1a93a8dfd5249b92e"></script> <script defer src="/assets/js/copy_code.js?12775fdf7f95e901d7119054556e495f" type="text/javascript"></script> <script defer src="/assets/js/jupyter_new_tab.js?d9f17b6adc2311cbabd747f4538bb15f"></script> <script async src="https://d1bxh8uas1mnw7.cloudfront.net/assets/embed.js"></script> <script async src="https://badge.dimensions.ai/badge.js"></script> <script defer type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script> <script src="/assets/js/mathjax-setup.js?a5bb4e6a542c546dd929b24b8b236dfd"></script> <script defer src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6" crossorigin="anonymous"></script> <script defer src="/assets/js/progress-bar.js?2f30e0e6801ea8f5036fa66e1ab0a71a" type="text/javascript"></script> <script src="/assets/js/vanilla-back-to-top.min.js?f40d453793ff4f64e238e420181a1d17"></script> <script>
    addBackToTop();
  </script> <script type="module" src="/assets/js/search/ninja-keys.min.js?a3446f084dcaecc5f75aa1757d087dcf"></script> <ninja-keys hidebreadcrumbs noautoloadmdicons placeholder="Type to start searching"></ninja-keys> <script src="/assets/js/search-setup.js?6c304f7b1992d4b60f7a07956e52f04a"></script> <script src="/assets/js/search-data.js"></script> <script src="/assets/js/shortcut-key.js?6f508d74becd347268a7f822bca7309d"></script> </body> </html>