<!DOCTYPE html> <html lang="en"> <head> <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"> <meta charset="utf-8"> <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"> <meta http-equiv="X-UA-Compatible" content="IE=edge"> <title> Mendelian randomization | HÃ©ctor Climente-GonzÃ¡lez </title> <meta name="author" content="HÃ©ctor Climente-GonzÃ¡lez"> <meta name="description" content="Causality."> <meta name="keywords" content="machine-learning, genetics, pharma, drug-discovery, climente, climente-gonzalez"> <link rel="stylesheet" href="/assets/css/bootstrap.min.css?a4b3f509e79c54a512b890d73235ef04"> <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/css/mdb.min.css" integrity="sha256-jpjYvU3G3N6nrrBwXJoVEYI/0zw8htfFnhT9ljN3JJw=" crossorigin="anonymous"> <link defer rel="stylesheet" href="/assets/css/academicons.min.css?f0b7046b84e425c55f3463ac249818f5"> <link defer rel="stylesheet" href="/assets/css/scholar-icons.css?62b2ac103a88034e6882a5be5f3e2772"> <link defer rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700|Roboto+Slab:100,300,400,500,700|Material+Icons&amp;display=swap"> <link defer rel="stylesheet" href="/assets/css/jekyll-pygments-themes-github.css?591dab5a4e56573bf4ef7fd332894c99" media="" id="highlight_theme_light"> <link defer href="/assets/css/bootstrap-toc.min.css?6f5af0bb9aab25d79b2448143cbeaa88" rel="stylesheet"> <link rel="shortcut icon" href="data:image/svg+xml,&lt;svg%20xmlns=%22http://www.w3.org/2000/svg%22%20viewBox=%220%200%20100%20100%22&gt;&lt;text%20y=%22.9em%22%20font-size=%2290%22&gt;%F0%9F%AA%A2&lt;/text&gt;&lt;/svg&gt;"> <link rel="stylesheet" href="/assets/css/main.css?d41d8cd98f00b204e9800998ecf8427e"> <link rel="canonical" href="https://hclimente.github.io/blog/mendelian-randomization/"> <script src="/assets/js/theme.js?9a0c749ec5240d9cda97bc72359a72c0"></script> <link defer rel="stylesheet" href="/assets/css/jekyll-pygments-themes-native.css?5847e5ed4a4568527aa6cfab446049ca" media="none" id="highlight_theme_dark"> <script>
    initTheme();
  </script> </head> <body class="fixed-top-nav sticky-bottom-footer"> <header> <nav id="navbar" class="navbar navbar-light navbar-expand-sm fixed-top" role="navigation"> <div class="container"> <a class="navbar-brand title font-weight-lighter" href="/"> <span class="font-weight-bold">HÃ©ctor</span> Climente-GonzÃ¡lez </a> <button class="navbar-toggler collapsed ml-auto" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar top-bar"></span> <span class="icon-bar middle-bar"></span> <span class="icon-bar bottom-bar"></span> </button> <div class="collapse navbar-collapse text-right" id="navbarNav"> <ul class="navbar-nav ml-auto flex-nowrap"> <li class="nav-item "> <a class="nav-link" href="/">about </a> </li> <li class="nav-item active"> <a class="nav-link" href="/blog/">blog </a> </li> <li class="nav-item "> <a class="nav-link" href="/publications/">publications </a> </li> <li class="nav-item "> <a class="nav-link" href="/repositories/">repositories </a> </li> <li class="nav-item "> <a class="nav-link" href="/cv/">cv </a> </li> <li class="nav-item"> <button id="search-toggle" title="Search" onclick="openSearchModal()"> <span class="nav-link">ctrl k <i class="ti ti-search"></i></span> </button> </li> <li class="toggle-container"> <button id="light-toggle" title="Change theme"> <i class="ti ti-sun-moon" id="light-toggle-system"></i> <i class="ti ti-moon-filled" id="light-toggle-dark"></i> <i class="ti ti-sun-filled" id="light-toggle-light"></i> </button> </li> </ul> </div> </div> </nav> <progress id="progress" value="0"> <div class="progress-container"> <span class="progress-bar"></span> </div> </progress> </header> <div class="container mt-5" role="main"> <div class="row"> <div class="col-sm-3"> <nav id="toc-sidebar" class="sticky-top"></nav> </div> <div class="col-sm-9"> <div class="post"> <header class="post-header"> <h1 class="post-title">Mendelian randomization</h1> <p class="post-meta"> Created in April 03, 2023 </p> <p class="post-tags"> <a href="/blog/2023"> <i class="fa-solid fa-calendar fa-sm"></i> 2023 </a> Â  Â· Â  <a href="/blog/tag/bioinformatics"> <i class="fa-solid fa-hashtag fa-sm"></i> bioinformatics</a> Â  <a href="/blog/tag/genetics"> <i class="fa-solid fa-hashtag fa-sm"></i> genetics</a> </p> </header> <article class="post-content"> <div id="markdown-content"> <h1 id="a-primer-on-mendelian-randomization">A primer on Mendelian randomization</h1> <p>Understanding how exposures, such as gene expression, cause complex phenotypes is a key question in biology. While randomized controlled trials are the gold standard for establishing causality, their use on humans is unethical, necessitating the use of observational studies. Crucially, observational studies cannot directly prove causality: they can produce biased estimates and be affected by unknown confounders and reverse causality. Interestingly, SNPs are generally equivalent to a randomized treatment, as they are randomized at conception and fixed throughout life, making them unaffected by confounders (though LD and population structure are exceptions to the former, and canalization to the latter). Hence, when this is true, we can indeed establish causality. This is the case between siblings, for which the inheritance of each SNP is completely randomized. However, in the general population, the distribution of variants is not completely random. For instance, due to assortative mating, people with similar heritable phenotypes will tend to get together. In general, traits that are more genetically proximal (e.g., molecular traits like gene expression) are less prone to these biases.</p> <blockquote> <p>ðŸ’¡ <strong>Randomized controlled trials</strong> (RCTs) can establish causality even in the presence of variables that are not under experimental control. The key procedure involves randomly assigning the treatment to the subjects. In other words, with a large enough sample size, the treatment and control groups should be similar in all regards except for the treatment variable. Hence, any difference between the groups must be attributable to the treatment.</p> </blockquote> <p>Mendelian Randomization (MR) studies the causal effect of an exposure on an outcome using genetic variants (SNPs). When certain assumptions are met (<a href="#key-assumptions-and-how-to-verify-them">Key assumptions and how to verify them</a>), it can address the aforementioned issues. MR leverages <a href="https://www.cell.com/cms/attachment/8c0c3e2d-595c-4c71-8562-85225737be90/gr2.jpg" rel="external nofollow noopener" target="_blank">this</a> <a href="https://mr-dictionary.mrcieu.ac.uk/term/dag/" rel="external nofollow noopener" target="_blank">directed acyclic graph</a> (DAG).</p> <h2 id="measuring-causal-effects-using-instruments">Measuring causal effects using instruments</h2> <p>Our goal is to precisely estimate the effect of an exposure on a trait. However, the presence of confounders that are either unobserved or difficult to measure makes this impossible. An instrument, such as the SNP, is a variable that modifies the exposure and is not affected by the confounders. Since there are no backdoor paths between the SNP and the outcome, any effect between them must occur through the exposure.</p> <p>In mathematical terms, we aim to estimate the effect of an exposure \(X\) on an outcome \(Y\)using an instrument \(Z\). Since \(Z\) is a SNP, it will usually take values in 0, 1 and 2 (the number of minor alleles). Assume some potential confounders affecting both \(X\) and \(Y\). The most basic MR protocol assumes linearity, and consists of three steps:</p> <ol> <li>Estimate the relation between the instrument and the exposure: \(X=\beta_{ZX}Z+\varepsilon\)</li> <li>Compute an estimated exposure, \(\tilde X = \beta_{ZX}Z\), which is independent of potential confounders</li> <li>Estimate \(Y=\beta_{\hat{X}Y}\hat{X}+\epsilon\)</li> </ol> <p>The estimated causal effect \(\beta_{\hat{X}Y}\) is unconfounded! This is usually performed via <a href="https://mr-dictionary.mrcieu.ac.uk/term/tsls/" rel="external nofollow noopener" target="_blank"><strong>two-stage least squares</strong></a>. Alternatively, we can obtain \(\beta_{\hat{X}Y}\) without computing it explicitly via the <strong><a href="https://mr-dictionary.mrcieu.ac.uk/term/wald-ratio/" rel="external nofollow noopener" target="_blank">Wald ratio estimator</a></strong>: \(\beta_{\hat{X}Y} = \frac {\beta_{ZY}} {\beta_{ZX}}\). Note that the standard error of \(\beta_{\hat{X}Y}\) also needs to be computed.</p> <h2 id="key-assumptions-and-how-to-verify-them">Key assumptions and how to verify them</h2> <p>MR relies on three assumptions:</p> <ol> <li> <em>Relevance</em>: the SNP is associated with the exposure.</li> <li> <em>Independence</em>: the SNP is not associated to the outcome through a confounder, measured or not.</li> <li> <em>Exclusion restriction</em>: the SNP is exclusively associated to the outcome through the exposure.</li> </ol> <p>Other common, but optional, assumptions are <em>linearity</em> and <em>homogeneity</em> (the effect does not vary across strata in the population, like SNP levels or sex).</p> <p>Although these assumptions are sometimes violated, we have ways to detect such cases:</p> <ol> <li>It can be violated when the SNP is a weak instrument. It can be measured using (Cragg-Donald) F-statistic \(\left( \frac {n-m-1} {m} \right) \left( \frac {R^2} {1-R^2}\right)\) where \(n\) is the number of samples, \(m\) is the number of SNPs, and \(R^2\) is the SNP heritability. Conventionally, this assumption is considered met when F is larger than 10.</li> <li>Common violations include <a href="https://mr-dictionary.mrcieu.ac.uk/term/assortative-mating/" rel="external nofollow noopener" target="_blank">assortative mating</a>, <a href="https://mr-dictionary.mrcieu.ac.uk/term/pop-strat/" rel="external nofollow noopener" target="_blank">population structure</a> and <a href="https://mr-dictionary.mrcieu.ac.uk/term/dynastic/" rel="external nofollow noopener" target="_blank">dynastic effects</a>. This can only be assessed for the potential confounders that have been measured, by measuring the respective associations between those and the SNP and outcome.</li> <li> <a href="https://mr-dictionary.mrcieu.ac.uk/term/horizontal-pleiotropy/" rel="external nofollow noopener" target="_blank">(Horizontal) pleiotropy</a> and linkage with other causal genes are common violations. When using multiple instruments, it can be assumed that they will all lead to the same estimate of the causal effect. Strong departures from this situation suggest that some instruments are affecting the outcome in different ways.</li> </ol> <h2 id="types-of-mr-algorithms">Types of MR algorithms</h2> <p>MR methods can be classified according to multiple criteria:</p> <ul> <li>Origin of the measurements <ul> <li>One-sample: the SNP, exposure and outcome are measured on the same samples. Example: two-stage least squares.</li> <li>Two-sample: the SNP-exposure relationship is measured on a set of samples, and the exposure-outcome on another one. Example: Wald ratio estimate</li> </ul> </li> <li>Instrument <ul> <li>SNP</li> <li>Polygenic score</li> </ul> </li> <li>Granularity <ul> <li>Individual-level</li> <li>Summary-level: as an attention note, all summary statistics must come from ancestry-matched samples.</li> </ul> </li> <li>Number of SNPs <ul> <li>Single SNP</li> <li>Multiple SNPs: this can address the weak association that individual SNPs might have with the trait. Notably, the dependencies between the SNPs (linkage disequilibrium) can produce trouble. The SNPs should be conditionally independent predictors (e.g., not in perfect LD). Even when not in perfect LD, too many correlated SNPs can lead to numerically unstable estimates. This is solvable by a prior feature selection and/or dimensionality reduction steps.</li> </ul> </li> <li>Number of exposures <ul> <li>Univariable: one exposure</li> <li>Multivariable: several exposures</li> </ul> </li> <li>Location of the SNPs <ul> <li>Cis-MR: use SNP(s) from the gene of interest coding region</li> <li>Trans-MR</li> </ul> </li> </ul> <h2 id="applications">Applications</h2> <h3 id="learning-the-causes-of-a-trait">Learning the causes of a trait</h3> <p>MR provides an ethical way to find causal links between exposures and diseases.</p> <p><strong>Exposure:</strong> it can be simple (e.g., gene expression), or complex (e.g., body mas index).</p> <p><strong>Instruments:</strong> we will use as many SNPs associated to the exposure of interest as possible (e.g., polygenic score). This will capture the full genetic architecture of the trait as well as violations of the assumptions.</p> <h3 id="finding-drug-targets">Finding drug targets</h3> <p>In the clinical setting, our interest is learning about *modifiable <strong>*</strong>causes of disease, i.e., those that can be treated.</p> <p><strong>Exposure:</strong> typically, a protein, since it needs to be targetable by a small molecule, a monoclonal antibody, etc. The expectation is that it will affect a complex biomarker through <em>vertical</em> pleiotropy (e.g., drugs that lower blood LDL do not target LDL directly, but hamper LDL/cholesterol synthesis).</p> <p><strong>Instruments:</strong> SNPs affecting the gene/protein of interest, often <em>cis</em>-pQTL. <em>Trans</em>-pQTL can also be used, but they are more likely to partake in horizontal pleiotropy.</p> <h3 id="identifying-the-relevant-tissues">Identifying the relevant tissues</h3> <p>We can use multivariable MR to get tissue-specific exposures, and partition the effect on the phenotype. Two disclaimers are in order: GWAS hits often colocalize with eQTLs in multiple tissues; and expression control in the diseased state might not match that in the healthy state.</p> <h3 id="characterizing-causal-pathways">Characterizing causal pathways</h3> <p>Multivariable MR studies the causal impact of multiple exposures. This allows to consider complex causal pathways, even those involving mediators.</p> <h1 id="references">References</h1> <ul> <li>Hartley, A. E., Power, G. M., Sanderson, E., &amp; Smith, G. D. (2022). A Guide for Understanding and Designing Mendelian Randomization Studies in the Musculoskeletal Field. In JBMR Plus (Vol. 6, Issue 10). Wiley. https://doi.org/10.1002/jbm4.10675</li> <li>Burgess, S., Mason, A. M., Grant, A. J., Slob, E. A. W., Gkatzionis, A., Zuber, V., Patel, A., Tian, H., Liu, C., Haynes, W. G., Hovingh, G. K., Knudsen, L. B., Whittaker, J. C., &amp; Gill, D. (2023). Using genetic association data to guide drug discovery and development: Review of methods and applications. In The American Journal of Human Genetics (Vol. 110, Issue 2, pp. 195â€“214). Elsevier BV. https://doi.org/10.1016/j.ajhg.2022.12.017</li> <li>Teumer, A. (2018). Common Methods for Performing Mendelian Randomization. In Frontiers in Cardiovascular Medicine (Vol. 5). Frontiers Media SA. https://doi.org/10.3389/fcvm.2018.00051</li> <li>Holmes, M. V., Richardson, T. G., Ference, B. A., Davies, N. M., &amp; Davey Smith, G. (2021). Integrating genomics with biomarkers and therapeutic targets to invigorate cardiovascular drug development. In Nature Reviews Cardiology (Vol. 18, Issue 6, pp. 435â€“453). Springer Science and Business Media LLC. https://doi.org/10.1038/s41569-020-00493-1</li> </ul> </div> </article> <div id="giscus_thread" style="max-width: 930px; margin: 0 auto;"> <script>
      let giscusTheme = determineComputedTheme();
      let giscusAttributes = {
        src: 'https://giscus.app/client.js',
        'data-repo': 'hclimente/hclimente.github.io',
        'data-repo-id': '',
        'data-category': 'Comments',
        'data-category-id': '',
        'data-mapping': 'title',
        'data-strict': '1',
        'data-reactions-enabled': '1',
        'data-emit-metadata': '0',
        'data-input-position': 'bottom',
        'data-theme': giscusTheme,
        'data-lang': 'en',
        crossorigin: 'anonymous',
        async: '',
      };

      let giscusScript = document.createElement('script');
      Object.entries(giscusAttributes).forEach(([key, value]) => giscusScript.setAttribute(key, value));
      document.getElementById('giscus_thread').appendChild(giscusScript);
    </script> <noscript>Please enable JavaScript to view the <a href="http://giscus.app/?ref_noscript" rel="external nofollow noopener" target="_blank">comments powered by giscus.</a> </noscript> </div> </div> </div> </div> </div> <footer class="sticky-bottom mt-5" role="contentinfo"> <div class="container"> Â© Copyright 2025 HÃ©ctor Climente-GonzÃ¡lez. </div> </footer> <script src="https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js" integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script> <script src="/assets/js/bootstrap.bundle.min.js"></script> <script src="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/js/mdb.min.js" integrity="sha256-NdbiivsvWt7VYCt6hYNT3h/th9vSTL4EDWeGs5SN3DA=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/masonry-layout@4.2.2/dist/masonry.pkgd.min.js" integrity="sha256-Nn1q/fx0H7SNLZMQ5Hw5JLaTRZp0yILA/FRexe19VdI=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/imagesloaded@5.0.0/imagesloaded.pkgd.min.js" integrity="sha256-htrLFfZJ6v5udOG+3kNLINIKh2gvoKqwEhHYfTTMICc=" crossorigin="anonymous"></script> <script defer src="/assets/js/masonry.js?a0db7e5d5c70cc3252b3138b0c91dcaf" type="text/javascript"></script> <script defer src="https://cdn.jsdelivr.net/npm/medium-zoom@1.1.0/dist/medium-zoom.min.js" integrity="sha256-ZgMyDAIYDYGxbcpJcfUnYwNevG/xi9OHKaR/8GK+jWc=" crossorigin="anonymous"></script> <script defer src="/assets/js/zoom.js?85ddb88934d28b74e78031fd54cf8308"></script> <script defer src="/assets/js/bootstrap-toc.min.js?c82ff4de8b0955d6ff14f5b05eed7eb6"></script> <script src="/assets/js/no_defer.js?2781658a0a2b13ed609542042a859126"></script> <script defer src="/assets/js/common.js?e0514a05c5c95ac1a93a8dfd5249b92e"></script> <script defer src="/assets/js/copy_code.js?12775fdf7f95e901d7119054556e495f" type="text/javascript"></script> <script defer src="/assets/js/jupyter_new_tab.js?d9f17b6adc2311cbabd747f4538bb15f"></script> <script async src="https://d1bxh8uas1mnw7.cloudfront.net/assets/embed.js"></script> <script async src="https://badge.dimensions.ai/badge.js"></script> <script defer type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script> <script src="/assets/js/mathjax-setup.js?a5bb4e6a542c546dd929b24b8b236dfd"></script> <script defer src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6" crossorigin="anonymous"></script> <script defer src="/assets/js/progress-bar.js?2f30e0e6801ea8f5036fa66e1ab0a71a" type="text/javascript"></script> <script src="/assets/js/vanilla-back-to-top.min.js?f40d453793ff4f64e238e420181a1d17"></script> <script>
    addBackToTop();
  </script> <script type="module" src="/assets/js/search/ninja-keys.min.js?a3446f084dcaecc5f75aa1757d087dcf"></script> <ninja-keys hidebreadcrumbs noautoloadmdicons placeholder="Type to start searching"></ninja-keys> <script src="/assets/js/search-setup.js?6c304f7b1992d4b60f7a07956e52f04a"></script> <script src="/assets/js/search-data.js"></script> <script src="/assets/js/shortcut-key.js?6f508d74becd347268a7f822bca7309d"></script> </body> </html>