<!DOCTYPE html> <html lang="en"> <head> <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"> <meta charset="utf-8"> <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"> <meta http-equiv="X-UA-Compatible" content="IE=edge"> <title> Mendelian randomization | Héctor Climente-González </title> <meta name="author" content="Héctor Climente-González"> <meta name="description" content="Causality from observations, thanks to genetics"> <meta name="keywords" content="machine-learning, genetics, pharma, drug-discovery, climente, climente-gonzalez"> <link rel="stylesheet" href="/assets/css/bootstrap.min.css?a4b3f509e79c54a512b890d73235ef04"> <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/css/mdb.min.css" integrity="sha256-jpjYvU3G3N6nrrBwXJoVEYI/0zw8htfFnhT9ljN3JJw=" crossorigin="anonymous"> <link defer rel="stylesheet" href="/assets/css/academicons.min.css?f0b7046b84e425c55f3463ac249818f5"> <link defer rel="stylesheet" href="/assets/css/scholar-icons.css?62b2ac103a88034e6882a5be5f3e2772"> <link defer rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700|Roboto+Slab:100,300,400,500,700|Material+Icons&amp;display=swap"> <link defer rel="stylesheet" href="/assets/css/jekyll-pygments-themes-github.css?591dab5a4e56573bf4ef7fd332894c99" media="" id="highlight_theme_light"> <link rel="shortcut icon" href="data:image/svg+xml,&lt;svg%20xmlns=%22http://www.w3.org/2000/svg%22%20viewBox=%220%200%20100%20100%22&gt;&lt;text%20y=%22.9em%22%20font-size=%2290%22&gt;%F0%9F%AA%A2&lt;/text&gt;&lt;/svg&gt;"> <link rel="stylesheet" href="/assets/css/main.css?d41d8cd98f00b204e9800998ecf8427e"> <link rel="canonical" href="https://hclimente.github.io/blog/mendelian-randomization/"> <script src="/assets/js/theme.js?a81d82887dd692e91686b43de4542f18"></script> <link defer rel="stylesheet" href="/assets/css/jekyll-pygments-themes-native.css?5847e5ed4a4568527aa6cfab446049ca" media="none" id="highlight_theme_dark"> <script>
    initTheme();
  </script> </head> <body class="fixed-top-nav sticky-bottom-footer"> <header> <nav id="navbar" class="navbar navbar-light navbar-expand-sm fixed-top" role="navigation"> <div class="container"> <a class="navbar-brand title font-weight-lighter" href="/"> <span class="font-weight-bold">Héctor</span> Climente-González </a> <button class="navbar-toggler collapsed ml-auto" type="button" data-toggle="collapse" data-target="#navbarNav" aria-controls="navbarNav" aria-expanded="false" aria-label="Toggle navigation"> <span class="sr-only">Toggle navigation</span> <span class="icon-bar top-bar"></span> <span class="icon-bar middle-bar"></span> <span class="icon-bar bottom-bar"></span> </button> <div class="collapse navbar-collapse text-right" id="navbarNav"> <ul class="navbar-nav ml-auto flex-nowrap"> <li class="nav-item "> <a class="nav-link" href="/">about </a> </li> <li class="nav-item active"> <a class="nav-link" href="/blog/">blog </a> </li> <li class="nav-item "> <a class="nav-link" href="/publications/">publications </a> </li> <li class="nav-item "> <a class="nav-link" href="/repositories/">repositories </a> </li> <li class="nav-item "> <a class="nav-link" href="/cv/">cv </a> </li> <li class="nav-item"> <button id="search-toggle" title="Search" onclick="openSearchModal()"> <span class="nav-link">ctrl k <i class="ti ti-search"></i></span> </button> </li> <li class="toggle-container"> <button id="light-toggle" title="Change theme"> <i class="ti ti-sun-moon" id="light-toggle-system"></i> <i class="ti ti-moon-filled" id="light-toggle-dark"></i> <i class="ti ti-sun-filled" id="light-toggle-light"></i> </button> </li> </ul> </div> </div> </nav> <progress id="progress" value="0"> <div class="progress-container"> <span class="progress-bar"></span> </div> </progress> </header> <div class="container mt-5" role="main"> <div class="post"> <header class="post-header"> <h1 class="post-title">Mendelian randomization</h1> <p class="post-meta"> Created on April 03, 2023 </p> <p class="post-tags"> <a href="/blog/2023"> <i class="fa-solid fa-calendar fa-sm"></i> 2023 </a>   ·   <a href="/blog/tag/bioinformatics"> <i class="fa-solid fa-hashtag fa-sm"></i> bioinformatics</a>   <a href="/blog/tag/genetics"> <i class="fa-solid fa-hashtag fa-sm"></i> genetics</a> </p> </header> <article class="post-content"> <div id="markdown-content"> <p>Randomized controlled trials (RCTs) are the gold standard to establish if an intervention <strong>causes</strong> an outcome. The magic lies in randomly assigning subjects to either the intervention or control group. A sufficiently large sample ensures any other confounding variables are evenly distributed. Thus, observed differences must be due to the intervention, and not to any other variables not under experimental control.</p> <p>A key question in biology is whether and how an exposure, like the protein expression of a gene, causes a complex phenotype. However, using RCTs to study this issue would be quite challenging technically—and ethically abominable. Hence, we need to make do with <em>observational</em> studies. Crucially, observational studies cannot directly prove causality, since they are plagued by biases, unknown confounders and reverse causality.</p> <h1 id="enter-mendelian-randomization">Enter Mendelian Randomization</h1> <p>Interestingly, <a href="https://en.wikipedia.org/wiki/Single-nucleotide_polymorphism" rel="external nofollow noopener" target="_blank">genetic variants</a> are akin to a randomized treatment when comparing siblings. Which allele each sibling gets is randomized at conception and fixed throughout life, making them unaffected by confounders. While there are exceptions (<a href="https://en.wikipedia.org/wiki/Linkage_disequilibrium" rel="external nofollow noopener" target="_blank">linkage disequilibrium</a> and <a href="https://en.wikipedia.org/wiki/Population_structure_(genetics)" rel="external nofollow noopener" target="_blank">population structure</a> to the former, and <a href="https://en.wikipedia.org/w/index.php?title=Canalization_(genetics)" rel="external nofollow noopener" target="_blank">canalization</a> to the latter), when this assumption holds, we can indeed estimate causality. However, many observational studies are carried out in the general population, where the distribution of variants is not completely randomized. For instance, due to assortative mating, people with similar heritable phenotypes will tend to… mate. However, even in the general population and as long as <a href="#key-assumptions-and-how-to-verify-them">certain assumptions are met</a> the distribution of SNPs mimics a random assignment. Mendelian Randomization (MR) exploits this fact to estimate the causal effect of an exposure on an outcome using genetic variants, often SNPs.</p> <p>Remember: our goal is to precisely estimate the effect of an exposure on a trait. However, the presence of confounders that are either unobserved or difficult to measure makes this impossible. An instrument, such as a SNP in MR, is our way out of this: it is a variable that modifies the exposure and is not affected by the confounders. Or, more graphically:</p> <pre><code class="language-mermaid">---
config:
    layout: elk
    look: handDrawn
---
graph TD
    Z[Instrument] --&gt; X[Exposure];
    X --&gt; Y[Trait];
    U[Confounder] --&gt; X;
    U --&gt; Y;

    Z --&gt;|❌| U;
    Z --&gt;|❌| Y;

    style Z fill:#D4EDDA,stroke:#28A745,stroke-width:2px,rx:8px,ry:8px;
    style X fill:#CCE5FF,stroke:#007BFF,stroke-width:2px,rx:8px,ry:8px;
    style Y fill:#FFF3CD,stroke:#FFC107,stroke-width:2px,rx:8px,ry:8px;
    style U fill:#F8D7DA,stroke:#DC3545,stroke-width:2px,stroke-dasharray: 5 5;

    %% Dashed lines for the "no direct effect"
    linkStyle 4 stroke-dasharray: 5 5;
    linkStyle 5 stroke-dasharray: 5 5;
</code></pre> <p>Since there are no backdoor paths between the SNP and the outcome, any effect between them must occur through the exposure.</p> <p>More formally, we aim to estimate the effect of an exposure \(X\) on an outcome \(Y\) using an instrument \(Z\). \(Z\) is a SNP. When SNPs are di-allelic, and since our focus is (diploid) humans, they can be encoded as a numerical variable containing the number of minor alleles of an individual: 0, 1 or 2. We’ll also assume some potential confounders affecting both \(X\) and \(Y\). The most basic MR protocol assumes linearity, and consists of three steps:</p> <ol> <li>Estimate the relationship between the instrument and the exposure by solving \(X = \beta_{ZX}Z + \varepsilon\) via a <a href="https://en.wikipedia.org/wiki/Genome-wide_association_study" rel="external nofollow noopener" target="_blank">GWAS</a>. For instance, if the exposure is gene expression, we will conduct an eQTL study.</li> <li>Use that model to estimate the exposure independent of potential confounders \(\tilde X = \beta_{ZX}Z\)</li> <li>Estimate the relationship between this adjusted exposure and the outcome by solving \(Y = \beta_{\hat{X}Y} \hat{X} + \epsilon\)</li> </ol> <p>The estimated causal effect, \(\beta_{\hat{X}Y}\), is unconfounded!</p> <p>Mathematically, this is usually performed via <a href="https://mr-dictionary.mrcieu.ac.uk/term/tsls/" rel="external nofollow noopener" target="_blank">two-stage least squares</a>. Alternatively, we can obtain \(\beta_{\hat{X}Y}\) without computing it explicitly via the <a href="https://mr-dictionary.mrcieu.ac.uk/term/wald-ratio/" rel="external nofollow noopener" target="_blank">Wald ratio estimator</a>: \(\beta_{\hat{X}Y} = \frac {\beta_{ZY}} {\beta_{ZX}}\). Note that the standard error of \(\beta_{\hat{X}Y}\) also needs to be computed.</p> <h1 id="key-assumptions-and-how-to-verify-them">Key assumptions and how to verify them</h1> <p>MR relies on three assumptions (<em>relevance</em>, <em>independence</em> and <em>exclusion restriction</em>), described below. Other common, but optional, assumptions are <em>linearity</em> and <em>homogeneity</em> (the effect does not vary across strata in the population, like SNP levels or sex).</p> <h2 id="relevance">Relevance</h2> <p>The instrument is <em>strongly</em> associated with the exposure. When this assumption is violated, we say that the instrument is a weak instrument. The strength of this association can be measured using (Cragg-Donald) F-statistic:</p> \[\left( \frac {n-m-1} {m} \right) \left( \frac {R^2} {1-R^2}\right)\] <p>where \(n\) is the number of samples, \(m\) is the number of instruments, and \(R^2\) is the instrument heritability. Conventionally, an instrument is considered <em>relevant</em> when \(F &gt; 10\).</p> <h2 id="independence">Independence</h2> <p>The instrument is not associated with the outcome through a confounder, measured or not. Common violations include <a href="https://mr-dictionary.mrcieu.ac.uk/term/assortative-mating/" rel="external nofollow noopener" target="_blank">assortative mating</a>, <a href="https://mr-dictionary.mrcieu.ac.uk/term/pop-strat/" rel="external nofollow noopener" target="_blank">population structure</a> and <a href="https://mr-dictionary.mrcieu.ac.uk/term/dynastic/" rel="external nofollow noopener" target="_blank">dynastic effects</a>. This is critical and it can only be assessed for the potential confounders that have been measured or can be estimated, by measuring the respective associations between those and the instrument and outcome.</p> <h2 id="exclusion-restriction">Exclusion restriction</h2> <p>The instrument is <em>exclusively</em> associated to the outcome through the exposure. <a href="https://mr-dictionary.mrcieu.ac.uk/term/horizontal-pleiotropy/" rel="external nofollow noopener" target="_blank">(Horizontal) pleiotropy</a> and linkage with other causal genes are common violations. When using multiple instruments, it can be assumed that they will all lead to the same estimate of the causal effect. Strong departures from this situation suggest that some instruments are affecting the outcome in different ways.</p> <h1 id="flavors-of-mr">Flavors of MR</h1> <p><a href="#enter-mendelian-randomization">We just saw</a> the simplest version of MR; but there are many algorithms to carry it out, depending on what data we have access to. I classify them here according to multiple criteria.</p> <h2 id="dataset">Dataset</h2> <p>We can conduct MR with varying <strong>numbers of datasets</strong>. In <em>one-sample MR</em>, the instrument, exposure and outcome are measured on the same subjects. This is the case of the <a href="#enter-mendelian-randomization">two-stage least squares</a> algorithm. In <em>two-sample MR</em>, the instrument-exposure relationship is measured on a set of samples, and the exposure-outcome on another one. One example is the Wald ratio estimate, which is the following statistic:</p> \[W = \frac {\beta_{ZY}} {\beta_{ZX}}\] <p>Typically, this is linked to the <strong>granularity</strong> of the data. One-sample MR is the ideal setting, since all the statistics are derived from the same dataset. This modality is easier when we have access to individual-level data. However, few datasets are rich enough to allow us to quantify effect sizes from the exposure and the outcome. A plethora of summary statistics are available in public repositories. For instance, we could measure the effect size of the instrument on gene expression on <a href="https://www.gtexportal.org" rel="external nofollow noopener" target="_blank">GTEx</a>, and on the outcome on a <a href="https://www.ebi.ac.uk/gwas/" rel="external nofollow noopener" target="_blank">publicly available GWAS</a>; this is called <em>two-sample MR</em>. As an attention note, all summary statistics need to come from ancestry-matched samples to preserve the <a href="#independence">independence</a> assumptions.</p> <h2 id="instrument">Instrument</h2> <p>The <strong>choice of instrument</strong> is a critical consideration. Instruments can range from a single SNP to multiple SNPs. A single SNP offers simplicity but is likely to have weak association with the exposure. While multiple SNPs can be a stronger instrument, they introduce challenges related to linkage disequilibrium and the need for conditional independence. In such a case, prior feature selection or dimensionality reduction techniques may be necessary to address these issues and ensure numerical stability. In the extreme case we could use a polygenic score (PGS), a weighted sum of multiple SNPs derived from a GWAS of the exposure, to capture a larger proportion of the variance in the exposure. Beware, though, of the potential biases lurking in the GWAS weights and the lingering effects of linkage disequilibrium.</p> <p>The <strong>location of the SNPs</strong> can be relevant when examining molecular traits like gene expression. In <em>cis-MR</em> we use SNPs located in the coding region of the target gene, or very nearby. This proximity dramatically boosts our confidence that the exclusion restriction assumption holds, as these SNPs are highly likely to affect gene expression directly. Alternatively, in <em>trans-MR</em> SNPs can be located very far from the exposure gene, and hence <a href="#exclusion-restriction">horizontal pleiotropy</a> can become an issue.</p> <h2 id="exposure">Exposure</h2> <p>While I have focused on the case in which we deal with a single exposure (<em>univariable MR</em>), there are cases in which we use <strong>multiple exposures</strong> (<em>Multivariable MR</em> or MVMR).</p> <p>There are multiple <strong>types of exposures</strong>. Some can be relatively simple phenotypes close to the genetics, like gene expression. Others can be complex, like the body mass index. There are also many intermediate cases, like a protein’s activity, or a biomarker that’s affected by such activity via <em>vertical pleiotropy</em>.</p> <h1 id="example-applications">Example applications</h1> <table> <thead> <tr> <th>Goal</th> <th>Exposure</th> <th>Instrument</th> <th>MR Flavor</th> </tr> </thead> <tbody> <tr> <td>Learn causes of a trait</td> <td>Simple or complex</td> <td>One or multiple exposure-associated SNPs</td> <td>Univariate MR</td> </tr> <tr> <td>Characterize causal pathways</td> <td>Multiple exposures, possibly with mediators</td> <td>Independent SNPs for all exposures</td> <td>MVMR to model joint/mediated effects</td> </tr> <tr> <td>Find drug targets</td> <td>Protein activity or biomarker</td> <td>cis-pQTLs (preferred) or trans-pQTLs</td> <td>Cis-MR (preferred) or trans-MR</td> </tr> <tr> <td>Identify relevant tissues</td> <td>Tissue-specific gene expression</td> <td>eQTLs stratified by tissue</td> <td>Tissue-Specific MVMR to partition effects</td> </tr> </tbody> </table> <h1 id="references">References</h1> <ul> <li>Hartley, A. E., Power, G. M., Sanderson, E., &amp; Smith, G. D. (2022). <a href="https://doi.org/10.1002/jbm4.10675" rel="external nofollow noopener" target="_blank">A Guide for Understanding and Designing Mendelian Randomization Studies in the Musculoskeletal Field</a>. JBMR Plus (Vol. 6, Issue 10).</li> <li>Burgess, S., Mason, A. M., Grant, A. J., Slob, E. A. W., Gkatzionis, A., Zuber, V., Patel, A., Tian, H., Liu, C., Haynes, W. G., Hovingh, G. K., Knudsen, L. B., Whittaker, J. C., &amp; Gill, D. (2023). <a href="https://doi.org/10.1016/j.ajhg.2022.12.017" rel="external nofollow noopener" target="_blank">Using genetic association data to guide drug discovery and development: Review of methods and applications</a>. The American Journal of Human Genetics (Vol. 110, Issue 2, pp. 195–214).</li> <li>Teumer, A. (2018). <a href="https://doi.org/10.3389/fcvm.2018.00051" rel="external nofollow noopener" target="_blank">Common Methods for Performing Mendelian Randomization</a>. Frontiers in Cardiovascular Medicine (Vol. 5).</li> <li>Holmes, M. V., Richardson, T. G., Ference, B. A., Davies, N. M., &amp; Davey Smith, G. (2021). <a href="https://doi.org/10.1038/s41569-020-00493-1" rel="external nofollow noopener" target="_blank">Integrating genomics with biomarkers and therapeutic targets to invigorate cardiovascular drug development</a>. Nature Reviews Cardiology (Vol. 18, Issue 6, pp. 435–453). Springer Science and Business Media LLC.</li> <li>Sanderson, E. (2021). <a href="https://doi.org/10.1101/cshperspect.a038984" rel="external nofollow noopener" target="_blank">Multivariable Mendelian Randomization and Mediation</a>. Cold Spring Harbor Perspectives in Medicine (Vol. 11, Issue 2).</li> </ul> </div> </article> <div id="giscus_thread" style="max-width: 930px; margin: 0 auto;"> <br> <script defer src="/assets/js/giscus-setup.js"></script> <noscript> Please enable JavaScript to view the <a href="http://giscus.app/?ref_noscript" rel="external nofollow noopener" target="_blank">comments powered by giscus.</a> </noscript> </div> </div> </div> <footer class="sticky-bottom mt-5" role="contentinfo"> <div class="container"> © Copyright 2025 Héctor Climente-González. </div> </footer> <script src="https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js" integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script> <script src="/assets/js/bootstrap.bundle.min.js"></script> <script src="https://cdn.jsdelivr.net/npm/mdbootstrap@4.20.0/js/mdb.min.js" integrity="sha256-NdbiivsvWt7VYCt6hYNT3h/th9vSTL4EDWeGs5SN3DA=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/masonry-layout@4.2.2/dist/masonry.pkgd.min.js" integrity="sha256-Nn1q/fx0H7SNLZMQ5Hw5JLaTRZp0yILA/FRexe19VdI=" crossorigin="anonymous"></script> <script defer src="https://cdn.jsdelivr.net/npm/imagesloaded@5.0.0/imagesloaded.pkgd.min.js" integrity="sha256-htrLFfZJ6v5udOG+3kNLINIKh2gvoKqwEhHYfTTMICc=" crossorigin="anonymous"></script> <script defer src="/assets/js/masonry.js?a0db7e5d5c70cc3252b3138b0c91dcaf" type="text/javascript"></script> <script defer src="https://cdn.jsdelivr.net/npm/mermaid@10.7.0/dist/mermaid.min.js" integrity="sha256-TtLOdUA8mstPoO6sGvHIGx2ceXrrX4KgIItO06XOn8A=" crossorigin="anonymous"></script> <script defer src="/assets/js/mermaid-setup.js?38ca0a0126f7328d2d9a46bad640931f" type="text/javascript"></script> <script defer src="https://cdn.jsdelivr.net/npm/medium-zoom@1.1.0/dist/medium-zoom.min.js" integrity="sha256-ZgMyDAIYDYGxbcpJcfUnYwNevG/xi9OHKaR/8GK+jWc=" crossorigin="anonymous"></script> <script defer src="/assets/js/zoom.js?85ddb88934d28b74e78031fd54cf8308"></script> <script src="/assets/js/no_defer.js?2781658a0a2b13ed609542042a859126"></script> <script defer src="/assets/js/common.js?e0514a05c5c95ac1a93a8dfd5249b92e"></script> <script defer src="/assets/js/copy_code.js?c8a01c11a92744d44b093fc3bda915df" type="text/javascript"></script> <script defer src="/assets/js/jupyter_new_tab.js?d9f17b6adc2311cbabd747f4538bb15f"></script> <script async src="https://d1bxh8uas1mnw7.cloudfront.net/assets/embed.js"></script> <script async src="https://badge.dimensions.ai/badge.js"></script> <script defer type="text/javascript" id="MathJax-script" src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script> <script src="/assets/js/mathjax-setup.js?a5bb4e6a542c546dd929b24b8b236dfd"></script> <script defer src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6" crossorigin="anonymous"></script> <script defer src="/assets/js/progress-bar.js?2f30e0e6801ea8f5036fa66e1ab0a71a" type="text/javascript"></script> <script src="/assets/js/vanilla-back-to-top.min.js?f40d453793ff4f64e238e420181a1d17"></script> <script>
    addBackToTop();
  </script> <script type="module" src="/assets/js/search/ninja-keys.min.js?a3446f084dcaecc5f75aa1757d087dcf"></script> <ninja-keys hidebreadcrumbs noautoloadmdicons placeholder="Type to start searching"></ninja-keys> <script src="/assets/js/search-setup.js?6c304f7b1992d4b60f7a07956e52f04a"></script> <script src="/assets/js/search-data.js"></script> <script src="/assets/js/shortcut-key.js?6f508d74becd347268a7f822bca7309d"></script> </body> </html>